File: 695.MaxAreaofIsland.go
func maxAreaOfIsland(grid [][]int) int {
    m, n := len(grid), len(grid[0])
    res := 0

    for i := 0; i < m; i++{
        for j := 0; j < n; j++{
            if grid[i][j] == 1{
                cnt := dfs(grid, i, j)
                if cnt > res{
                    res = cnt
                }
            }
        }
    }
    
    return res
}

func dfs(grid [][]int, r ,c int) int {
    if r < 0 || r >= len(grid) || c < 0 || c >= len(grid[0]) || grid[r][c] != 1{
        return 0
    }

    grid[r][c] = 0
    return 1 + dfs(grid, r + 1, c) + dfs(grid, r - 1, c) + dfs(grid, r, c + 1) + dfs(grid, r, c - 1)
} 

File: 695.MaxAreaofIsland.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0]) 
        res = 0
        
        def dfs(i, j):
            if i < 0 or i >= m or j < 0 or j >= n or grid[i][j] != 1:
                return 0
            
            grid[i][j] = 0
            return 1 + dfs(i + 1, j) + dfs(i, j + 1) + dfs(i - 1, j) + dfs(i, j - 1)

        for i in range(m):
            for j in range(n):
                cnt = 0
                if grid[i][j] == 1:
                    cnt = dfs(i, j)
                    res = max(res, cnt)

        return res
